src/cxy_ctl.erl

```diff
@@ -93,13 +95,24 @@ is_valid_limit(inline_only) -> true;
 is_valid_limit(_)           -> false.
 
 %% Locations of cxy counter tuple positions used for ets:update_counter
-%% {Task_Type, Stored_Max_Procs_Value, Active_Procs, Max_History, Slow_Factor_As_Percentage}
+%% {Task_Type, Stored_Max_Procs_Value, Active_Procs, Max_History, Slow_Factor_As_Percentage, High_Water_Procs}
 -define(MAX_PROCS_POS,    2).
 -define(ACTIVE_PROCS_POS, 3).
 -define(MAX_HISTORY_POS,  4).
-    
+-define(HIGH_WATER_PROCS_POS, 6).
+
 incr_active_procs(Task_Type) ->
-    ets:update_counter(?MODULE, Task_Type, {?ACTIVE_PROCS_POS, 1}).
+    [ Active_Procs, High_Water_Procs ] =
+        ets:update_counter(?MODULE, Task_Type, [{?ACTIVE_PROCS_POS, 1}, {?HIGH_WATER_PROCS_POS, 0}]),
+    if Active_Procs > -High_Water_Procs ->
+        %% Use ets:update_counter to implement max(High_Water_Procs,
+        %% Active_Procs).  We use the -Active_Procs because that allows us to
+        %% use a quirk of the ets:update_counter threshold interface to
+        %% atomically maintain the max.
+        ets:update_counter(?MODULE, Task_Type, {?HIGH_WATER_PROCS_POS, 0, -Active_Procs, -Active_Procs});
+       true -> ok
+    end,
+    Active_Procs.
```

Comment:
I never use 'if' in my code, I've banished it from the language since it carries too much baggage from other languages and doesn't operate the same way (and I've never found it useful). I write one-armed ifs using 'andalso' or 'orelse' so I would prefer to see this as:

```erlang
Active_Procs > -HIgh_Water_Procs
     andalso ets:update_counter(...),
Active_Procs.
```